File: zzippu_ios/zzippu/Data/Network/DTOs/GrowthDTO.swift
// Data/Network/DTOs/GrowthDTO.swift
// 서버 GrowthResponse / CreateGrowthRequest 대응

import Foundation

// MARK: - Response DTO

struct GrowthResponseDTO: Decodable {
    let id: UUID
    let babyId: UUID
    let recordedAt: String          // date(YYYY-MM-DD) — Mapper에서 변환
    let weightG: Int?
    let heightCm: Double?
    let headCircumferenceCm: Double?
    let temperatureC: Double?
    let memo: String?
    let createdAt: Date             // datetime — 자동 디코딩
}

// MARK: - Create Request DTO

struct GrowthCreateRequestDTO: Encodable {
    let recordedAt: String          // YYYY-MM-DD
    let weightG: Int?
    let heightCm: Double?
    let headCircumferenceCm: Double?
    let temperatureC: Double?
    let memo: String?
}

// MARK: - Update Request DTO (PATCH — 전체교체)

/// 서버 PATCH /growth/{id} 대응. recorded_at은 ISO datetime(apiEncoder가 Date→ISO 인코딩).
/// 나머지는 옵셔널이나 편집 시 전체 전송(전체교체 의미).
struct GrowthUpdateRequestDTO: Encodable {
    let recordedAt: Date            // datetime — apiEncoder가 ISO8601로 인코딩
    let weightG: Int?
    let heightCm: Double?
    let headCircumferenceCm: Double?
    let temperatureC: Double?
    let memo: String?
}

File: backend/app/application/dto/growth_dto.py
from dataclasses import dataclass
from datetime import date, datetime
from uuid import UUID


@dataclass
class CreateGrowthDTO:
    baby_id: UUID
    recorded_at: date
    weight_g: int | None = None
    height_cm: float | None = None
    head_circumference_cm: float | None = None
    memo: str | None = None
    id: UUID | None = None  # 클라 생성 UUID(멱등 upsert)


@dataclass
class UpdateGrowthDTO:
    # 전체 교체(PUT류). record_id로 조회 후 baby_id 소유권 검증에 사용.
    baby_id: UUID
    record_id: UUID
    recorded_at: date | None = None
    weight_g: int | None = None
    height_cm: float | None = None
    head_circumference_cm: float | None = None
    memo: str | None = None


@dataclass
class GrowthResponseDTO:
    id: UUID
    baby_id: UUID
    recorded_at: date
    weight_g: int | None
    height_cm: float | None
    head_circumference_cm: float | None
    memo: str | None
    created_at: datetime
